src/lib/components/Canvas.svelte

```diff
@@ -0,0 +1,157 @@
+<script context="module" lang="ts">
+  import RenderManager from '../util/renderManager';
+  import { getContext as getCTX } from 'svelte';
+
+  export const KEY = Symbol();
+
+  interface TypedContext {
+    register: RenderManager['register'];
+    unregister: RenderManager['unregister'];
+    redraw: RenderManager['redraw'];
+  }
+
+  export const getTypedContext = (): TypedContext => getCTX(KEY);
+</script>
+
+<script lang="ts">
+  import { onMount, onDestroy, setContext } from 'svelte';
+
+  export let width = 640,
+    height = 640,
+    pixelRatio: number | null = null,
+    style = '',
+    autoclear = true;
+
+  /** Class field. Only works for global classes. */
+  let clazz = '';
+
+  export { clazz as class, redraw, getCanvas, getContext };
+
+  let canvas: HTMLCanvasElement;
+  let context: CanvasRenderingContext2D | null = null;
+  let animationLoop: number;
+  let layerRef: HTMLDivElement;
+  let layerObserver: MutationObserver;
+
+  const manager = new RenderManager();
+
+  function redraw() {
+    manager.redraw();
+  }
+
+  function getCanvas() {
+    return canvas;
+  }
+
+  function getContext() {
+    return context;
+  }
+
+  if (pixelRatio === undefined) {
+    pixelRatio = 2;
+  }
+
+  function draw() {
+    if (!pixelRatio) return;
+    if (!context) return;
+    manager.render({ context, width, height, pixelRatio, autoclear });
+    animationLoop = window.requestAnimationFrame(draw);
+  }
+
+  setContext(KEY, {
+    register: manager.register,
+    unregister: manager.unregister,
+    redraw: manager.redraw
+  });
+
+  onMount(() => {
+    pixelRatio = window.devicePixelRatio;
+    const browserContext = canvas.getContext('2d');
+    if (!browserContext) {
+      throw Error("No 2D browser context found. This shouldn't happen.");
+    }
```

Comment:
Do we need this? As you say, it shouldn't happen and I can't think of any situation where it would

Reply:
([From Mozilla Docs](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext)):
> getContext() method returns a drawing context on the canvas, or null if the context identifier is not supported, or the canvas has already been set to a different context mode.

It seems that this check can just be replaced with a null assertion as '2d' has been compatible since canvas existed.